db/selector.go

```diff
@@ -0,0 +1,145 @@
+// gorp's DbMap.
+func (ts typeSelector[T]) SelectRowsFrom(ctx context.Context, tablename string, clauses string, args ...interface{}) (*typeRows[T], error) {
+	// Look up the table to use based on the type of this TypeSelector.
+	var throwaway T
+	t := reflect.TypeOf(throwaway)
+	tableMap, err := ts.wrapped.TableFor(t, false)
+	if err != nil {
+		return nil, fmt.Errorf("database model type not mapped to table name: %w", err)
+	}
+
+	// Extract the list of column names from the type's struct tags.
+	var columns []string
+	for _, column := range tableMap.Columns {
+		columns = append(columns, column.ColumnName)
+	}
+
+	// Iterate over the columns in the order they appear. For each, find the field
+	// on the struct type that has a matching `db:"colname"` struct tag. Save that
+	// field's index into a map for quick lookup later.
+	colIndexToFieldIndex := make([][]int, len(columns))
+	for i, column := range columns {
+		structField, found := t.FieldByNameFunc(func(fieldName string) bool {
+			structField, _ := t.FieldByName(fieldName)
+			tagColumn := structField.Tag.Get("db")
+			return tagColumn == column
+		})
+		if !found {
+			// This should never happen, as the columns were derived from the struct
+			// fields in the first place.
+			return nil, fmt.Errorf("no struct field with tag matching column %q", column)
+		}
+		colIndexToFieldIndex[i] = structField.Index
+	}
+
+	// Construct the query from the column names, table name, and given clauses.
+	query := fmt.Sprintf(
+		"SELECT %s FROM %s %s",
+		strings.Join(columns, ", "),
+		tablename,
+		clauses,
+	)
+
+	rows, err := ts.wrapped.WithContext(ctx).Query(query, args...)
+	if err != nil {
+		return nil, fmt.Errorf("reading db: %w", err)
+	}
+
+	return &typeRows[T]{wrapped: rows, idxMap: colIndexToFieldIndex}, nil
+}
+
+// typeRows is a wrapper around the stdlib's sql.typeRows, but with a more
+// type-safe method to get actual row content.
+type typeRows[T any] struct {
+	wrapped *sql.Rows
+	idxMap  [][]int
+}
+
+// Next is a wrapper around sql.Rows.Next(). It must be called before every call
+// to Get(), including the first.
+func (r typeRows[T]) Next() bool {
+	return r.wrapped.Next()
+}
+
+// Get is a wrapper around sql.Rows.Scan(). Rather than populating an arbitrary
+// number of &interface{} arguments, it returns a populated object of the
+// parameterized type.
+func (r typeRows[T]) Get() (*T, error) {
+	// We have to use reflect.New(reflect.TypeOf(var T)) to create our result
+	// value to ensure that it gets allocated on the heap and its individual
+	// fields are all addressable.
+	var throwaway T
+	t := reflect.TypeOf(throwaway)
+	v := reflect.New(t)
+
+	// Because sql.Rows.Scan(...) takes a variadic number of individual targets to
+	// read values into, build a slice that can be splatted into the call. Use the
+	// pre-computed map of column indices to field indices to populate it.
+	scanTargets := make([]interface{}, len(r.idxMap))
```

Comment:
Yeah, I think this would only be worthwhile if we were also eliminating the allocation to return an object, since it would allow this function to be free of heap allocations. But since we're allocating a value to return, we might as well do this allocation too. If profiling someday shows this to be a bottleneck (it probably won't be), we can always change it.